src/onnxflow/justbuildit/export.py

```diff
@@ -0,0 +1,614 @@
+     - state.inputs is a dict that represents valid kwargs to the forward
+        function of state.model
+
+    Outputs:
+     - A *-base.onnx file that implements state.model given state.inputs
+    """
+
+    def __init__(self):
+        super().__init__(
+            unique_name="export_pytorch",
+            monitor_message="Exporting PyTorch to ONNX",
+        )
+
+    def fire(self, state: build.State):
+        if not isinstance(state.model, (torch.nn.Module, torch.jit.ScriptModule)):
+            msg = f"""
+            The current stage (ExportPytorchModel) is only compatible with
+            models of type torch.nn.Module or torch.jit.ScriptModule, however
+            the stage received a model of type {type(state.model)}.
+            """
+            raise exp.StageError(msg)
+
+        # TODO: check that state.inputs is valid
+        # https://git.groq.io/code/Groq/-/issues/13947
+
+        # The `torch.onnx.export()` function accepts a tuple of positional inputs
+        # followed by a dictionary with all keyword inputs.
+        # The dictionary must be last item in tuple.
+        user_provided_args = list(state.inputs.keys())
+
+        if isinstance(state.model, torch.nn.Module):
+            # Validate user provided args
+            all_args = list(inspect.signature(state.model.forward).parameters.keys())
+
+            for inp in user_provided_args:
+                if inp not in all_args:
+                    msg = f"""
+                    Input name {inp} not found in the model's forward method. Available
+                    input names are: {all_args}"
+                    """
+                    raise ValueError(msg)
+
+            # Most pytorch models have args that are kind = positional_or_keyword.
+            # The `torch.onnx.export()` function accepts model args as
+            #     (all_positional_args_value,{keyword_arg:value}).
+            # To map the input_args correctly and to build an accurate model
+            # the order of the input_names must reflect the order of the model args.
+
+            # Collect order of pytorch model args.
+            all_args_order_mapping = {arg: idx for idx, arg in enumerate(all_args)}
+
+            # Sort the user provided inputs with respect to model args and store as tuple.
+            sorted_user_inputs = sorted(
+                user_provided_args, key=lambda x: all_args_order_mapping[x]
+            )
+            dummy_input_names = tuple(sorted_user_inputs)
+
+            # If a single input is provided torch.onnx.export will
+            # not accept a dictionary, so pop the first arg
+            user_args = copy.deepcopy(state.inputs)
+            first_input = user_args.pop(dummy_input_names[0])
+
+            # Create tuple: (first input, {rest of user_args dict as keyword args})
+            dummy_inputs = (first_input, user_args)
+
+        # TODO: Test with optional inputs
+        # https://git.groq.io/code/Groq/-/issues/14581
```

Comment:
* Remove link to GroqFlow issue
* Replace with MLAgility issue
* Cross reference: Mention MLAgility issue in the GroqFlow issue

Reply:
This issue is already closed!